src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -358,9 +388,14 @@ struct RemoveRecursiveObjectStorageOperation final : public IDiskObjectStorageOp
         else
         {
             for (auto it = metadata_storage.iterateDirectory(path_to_remove); it->isValid(); it->next())
+            {
+                if (local_paths_to_remove.size() == batch_size)
+                    return;
                 removeMetadataRecursive(tx, it->path());
-
-            tx->removeDirectory(path_to_remove);
+            }
+            /// Do not delete in case directory contains >= batch_size files
+            if (local_paths_to_remove.size() < batch_size)
```

Comment:
What if after processing of whole directory (no more files in it) `local_paths_to_remove.size() == batch_size`. 
Do we assume here that we will delete this directory on the next call of `execute` ?

Reply:
Yes, this case is also covered in tests